driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java

```diff
@@ -172,35 +176,32 @@ public void getAsync(final SingleResultCallback<InternalConnection> callback) {
                 errHandlingCallback.onResult(null, checkOutFailed(failure));
             }
         };
-        PooledConnection immediateConnection = null;
-
-        try {
-            immediateConnection = getPooledConnection(Timeout.immediate());
-        } catch (MongoTimeoutException e) {
-            // fall through
-        } catch (RuntimeException e) {
-            eventSendingCallback.onResult(null, e);
-            return;
-        }
-
-        if (immediateConnection != null) {
-            openAsync(immediateConnection, timeout, eventSendingCallback);
-        } else {
-            workers.getter().execute(() -> {
-                if (timeout.expired()) {
-                    eventSendingCallback.onResult(null, createTimeoutException(timeout));
-                    return;
-                }
+        asyncWorkManager.enqueue(new Task(timeout, t -> {
+            if (t != null) {
+                eventSendingCallback.onResult(null, t);
+            } else {
                 PooledConnection connection;
                 try {
                     connection = getPooledConnection(timeout);
                 } catch (RuntimeException e) {
                     eventSendingCallback.onResult(null, e);
                     return;
                 }
-                openAsync(connection, timeout, eventSendingCallback);
-            });
-        }
+                if (connection.opened()) {
+                    if (LOGGER.isTraceEnabled()) {
+                        LOGGER.trace(format("Pooled connection %s to server %s is already open",
+                                getId(connection), serverId));
+                    }
+                    eventSendingCallback.onResult(connection, null);
+                } else {
+                    if (LOGGER.isTraceEnabled()) {
+                        LOGGER.trace(format("Pooled connection %s to server %s is not yet open",
+                                getId(connection), serverId));
+                    }
+                    openConcurrencyLimiter.openAsyncOrGetAvailable(connection, timeout, eventSendingCallback);
```

Comment:
IIUC, this is ok to do in the same task because the timeout to wait for a permit will still always be shorter than the timeouts for all subsequent timeouts in the queue?

Reply:
All checkout tasks are added to the single queue with the same timeout `settings.getMaxWaitTime(NANOSECONDS)` started before they are added to the queue. Each task then leaves the queue to be handled by the `AsyncGetter` thread, and the thread always completes the task (even if it has to wait due to both/either `maxSize`/`maxConnecting` limits) before taking another task. Thus, each task (whether queued or being handled) has less time remaining until it times out than any task after it in the queue (of than any queued task if we are talking about the task that is being handled)<sup>1</sup>. Because of this, no queued task can time out while 

----

<sup>1</sup> This is a lie and a simplification, yet we are fine. In reality, because the pair of actions (start counting down time for a task, enqueue the task) is not done atomically with regard to other such pairs, it is possible that the aforementioned guarantee is violated. However, in order to realize that, one needs to have access to the queue, which users of `DefaultConcurrentPool` do not have. Not only this detail does not violate our best effort to not overdue timeouts, but also it is how things were in the `DefaultConcurrentPool` forever.

Reply:
I think I came up with a better explanation for the footnote 2:

1. Pool totally orders `getAsync` operations by putting the corresponding tasks in the queue.
2. For operations that are sequential (i.e. are ordered by `hb`), the queue order is consistent with `hb` and is also consistent with the order by `timeout.remaining`. For such operations a pool's user has an expectation that an earlier operation must timeout before the later operation.
3. For operations that are concurrent, the queue order may not be consistent with the order by `timeout.remaining`. For such operations a pool's user has no expectation on which must timeout first because they are not ordered from the user's perspective. So the fact that the pool may order them not by `timeout.remaining` is irrelevant to a user.

Reply:
Thanks for the detailed explanation!